### ISSUE TITLE
Some figures are stretched

### ISSUE THREAD

**Issue by michaeldorman** [geocompx/geocompy#230]
Some of the figures are stretched, for example:
https://py.geocompx.org/03-spatial-operations#fig-raster-slope
![image](https://github.com/geocompx/geocompy/assets/9386655/5a7592ab-8c78-4973-8afe-aeba117ae1b6)

Interestingly, they look fine in the local version:
![image](https://github.com/geocompx/geocompy/assets/9386655/2180f9c2-1c29-4b75-8134-e7f925c38ba5)

**Comment by Robinlovelace** [comment]
Weird, I have no idea what's causing that.

**Comment by Robinlovelace** [comment]
Stretched on my computer at https://py.geocompx.org/03-spatial-operations#fig-raster-slope also.

**Comment by michaeldorman** [comment]
P.S. The image files themselves are fine, but in the GitHub version for some reason the `<img>` `height` HTML property is being set to a specific value, therefore the images appear stretched:
![image](https://github.com/geocompx/geocompy/assets/9386655/0bbf7b9d-4d29-4687-ae8c-2f879708240a)

In the local version `height` isn't set so the images look fine:
![image](https://github.com/geocompx/geocompy/assets/9386655/8dd20e02-49f2-4855-8377-5698a30c7c6c)

So again my only idea is to re-install quarto or check it's settings on GitHub (don't know how to do that, will be happy if you can please help)

**Comment by Robinlovelace** [comment]
> Maybe the quarto version on GitHub is different? Just a thought

Yes, that could be it. Will update Quarto version and check again.

**Comment by Robinlovelace** [comment]
However, probably won't be until next week sorry, will asign myself as reminder.

**Comment by michaeldorman** [comment]
Thanks @Robinlovelace

**Comment by michaeldorman** [comment]
@Robinlovelace does it make sense to use quarto 1.3 on GitHub actions, at least as a temporary solution?

**Comment by Nowosad** [comment]
Quarto <1.5 has this bug btw -- https://github.com/geocompx/geocompy/issues/224#issuecomment-2122388111

**Comment by Robinlovelace** [comment]
Upstream fix: https://github.com/geocompx/docker/issues/48

**Comment by Robinlovelace** [comment]
Let's see if latest commit in geocompy fixes it...

**Comment by Robinlovelace** [comment]
Still doesn't seem to be fixed, sorry. Let's leave it a few more commits to ensure Docker and website is updated, we can fix in the geocompy .yml file if that doesn't do it.

**Comment by Robinlovelace** [comment]
Update on this: I think it *is* a Quarto bug, but that that bug persists after v1.5.41:

![image](https://github.com/geocompx/geocompy/assets/1825120/52e8e33c-728d-4df2-a029-9f9e6b3392ad)

That is reproducible from inside the latest version of the image defined in our .devcontainer file.

**Comment by Robinlovelace** [comment]
Optimistically closed with upstream commit, let's see if it actually fixes it!

**Comment by michaeldorman** [comment]
Hope so, thanks @Robinlovelace !

**Comment by Robinlovelace** [comment]
Minor update, can reproduce in devcontainer, going to try with latest version of quarto..

**Comment by Robinlovelace** [comment]
I think the best solution would be an upstream fix: https://github.com/quarto-dev/quarto-cli/issues/9927

**Comment by Robinlovelace** [comment]
Looks like there's a work-around documented by Quarto developer here: https://github.com/quarto-dev/quarto-cli/issues/10146

Can you give that a try if you get a chance @michaeldorman or anyone? It's odd because I thought we had captions and subcaps but not sure...

**Comment by michaeldorman** [comment]
> Looks like there's a work-around documented by Quarto developer here: [quarto-dev/quarto-cli#10146](https://github.com/quarto-dev/quarto-cli/issues/10146)
> 
> Can you give that a try if you get a chance @michaeldorman or anyone? It's odd because I thought we had captions and subcaps but not sure...

I don't really understand how to implement this fix and what's going on... 

The "good" example from the thread you mentioned does work fine standalone:
```py
#| label: fig-test
#| layout-ncol: 2
#| fig-subcap:
#|   - "&nbsp;"
#|   - "&nbsp;"
#| fig-cap: cap
import matplotlib.pyplot as plt
plt.plot([1,23,2,4])
plt.show()

plt.plot([8,65,23,90])
plt.show()
```
![image](https://github.com/geocompx/geocompy/assets/9386655/65c595c6-9c6d-45be-9d32-5e48ad6a1fdf)

When embedded in our book, it becomes stretched but with digits also stretched only in the local preview:

![image](https://github.com/geocompx/geocompy/assets/9386655/247032b1-4a68-434f-a524-c1b3191cbb85)

![image](https://github.com/geocompx/geocompy/assets/9386655/fdeae11f-1cf4-49d5-8a30-c613d4f3a013)

Anyway, the other figures in our book do contain `fig-subcap` like you say, and still they are stretched, so I don't see how to implement the solution even if it is indeed a solution. 

Also, using `plt.subplots(figsize=(x,y))` does affect figure size but the figures still come out stretched, so it doesn't solve the issue either.

**Comment by Robinlovelace** [comment]
Many thanks for testing Michael, I've passed that info on in the upstream issue. I cannot think of any other work-arounds at this point and am hopeful of an upstream fix :pray:

**Comment by michaeldorman** [comment]
Thanks @Robinlovelace 👍

**Comment by michaeldorman** [comment]
> Many thanks for testing Michael, I've passed that info on in the upstream issue. I cannot think of any other work-arounds at this point and am hopeful of an upstream fix 🙏

Maybe waiting for a quarto fix is a good idea. I just noticed that in the PDF version figures are not stretched, so I guess we can leave the HTML as-is for now

**Comment by DOSull** [comment]
Not sure where things stand with this, but it was the first thing I noticed in the mapping chapter where all the 3-panel NZ maps are stretched :(

**Comment by michaeldorman** [comment]
> Not sure where things stand with this, but it was the first thing I noticed in the mapping chapter where all the 3-panel NZ maps are stretched :(

Yes, unfortunately, you're right. I hope that this will be fixed in quarto, otherwise we can avoid using the multi-panel functionality of quarto (i.e., place all panels in the same image file), which will make the code in the book more complicated (after the book is printed). What do you think @Robinlovelace , @Nowosad , @anitagraser ?

**Comment by Robinlovelace** [comment]
This unmerged PR shows a work-around, a bit verbose but worked as far as I recall: https://github.com/geocompx/geocompy/pull/233

**Comment by Nowosad** [comment]
@michaeldorman @Robinlovelace @DOSull update: I think I've fixed the issue.

1. I used Robin's reprex -- it helped me discover that while the images are stretched in html, when you open them separately they have a correct aspect ratio
2. This suggested that the issue is in the css styling in quarto, not in the rendering 
3. I added a new tiny css file limiting the maximum width -- https://github.com/geocompx/geocompy/commit/aa9ec72a4fb4e0a9a04ea7a7c22a418bcfbf9348
4. Result:

![image](https://github.com/user-attachments/assets/0287923b-65a4-40e5-948a-40cce04d4194)
 

Please double-check if I have not destroyed anything else in the process...

**Comment by Robinlovelace** [comment]
Great fix @Nowosad sounds good to me. Indeed it is the CSS but the figures themselves. I wonder if your fix could be useful upstream in Quarto..

**Comment by DOSull** [comment]
Happy that my little nudge to revisit this has led to a solution! Good job team!
